src/main/java/org/elasticsearch/transport/netty/NettyTransport.java

```diff
@@ -426,6 +400,22 @@ public boolean onPortNumber(int portNumber) {
             throw new BindTransportException("Failed to bind to [" + port + "]", lastException.get());
         }
 
+        if (!DEFAULT_PROFILE.equals(name)) {
+            InetSocketAddress boundAddress = (InetSocketAddress) serverChannels.get(name).getLocalAddress();
+            InetSocketAddress publishAddress;
+            int publishPort = settings.getAsInt("publish_port", 0);
+            if (0 == publishPort) {
+                publishPort = boundAddress.getPort();
+            }
+            try {
+                String publishHost = settings.get("publish_host", this.settings.get("transport.publish_host", this.settings.get("transport.host")));
+                publishAddress = new InetSocketAddress(networkService.resolvePublishHostAddress(publishHost), publishPort);
+            } catch (Exception e) {
+                throw new BindTransportException("Failed to resolve publish address", e);
+            }
+            profileBoundAddresses.put(name, new BoundTransportAddress(new InetSocketTransportAddress(boundAddress), new InetSocketTransportAddress(publishAddress)));
+        }
```

Comment:
I got confused at first about this new block of code. It adds ability to configure publish host and port per profile, something that was missing before. I wonder if this change deserves its own specific PR, otherwise let's just update the description of this PR to clarify that this change is made too. Also, I wonder if we can reuse code that seems the same in [`doStart`](https://github.com/elasticsearch/elasticsearch/blob/master/src/main/java/org/elasticsearch/transport/netty/NettyTransport.java#L299)